pkg/storegateway/postings_codec.go

```diff
@@ -97,6 +105,141 @@ func diffVarintSnappyDecode(input []byte) (index.Postings, error) {
 	return newDiffVarintPostings(raw), nil
 }
 
+// isDiffVarintSnappyEncodedPostings returns true, if input looks like it has been encoded by diff+varint+snappy+matchers codec.
+func isDiffVarintSnappyWithmatchersEncodedPostings(input []byte) bool {
+	return bytes.HasPrefix(input, []byte(codecHeaderSnappyWithmatchers))
+}
+
+// diffVarintSnappyMatchersEncode encodes postings into snappy-encoded diff+varint representation,
+// prepended with any unapplied matchers to the result.
+// Returned byte slice starts with codecHeaderSnappyWithmatchers header.
+// Length argument is expected number of postings, used for preallocating buffer.
+func diffVarintSnappyMatchersEncode(p index.Postings, length int, unappliedMatchers []*labels.Matcher) ([]byte, error) {
+	varintPostings, err := diffVarintEncodeNoHeader(p, length)
+	if err != nil {
+		return nil, err
+	}
+
+	// Estimate sizes
+	matchersLen := encodedMatchersLen(unappliedMatchers)
```

Comment:
[nit] I think this makes it more clear.

```suggestion
	expectedMatchersLen := encodedMatchersLen(unappliedMatchers)
```